src/core/geometry/qgssfcgalgeometry.h

```diff
@@ -0,0 +1,516 @@
+#ifndef QGSSGCGAL_GEOMETRY_H
+#define QGSSGCGAL_GEOMETRY_H
+
+#define SIP_NO_FILE
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgsabstractgeometry.h"
+#include "qgscircularstring.h"
+#include "qgscompoundcurve.h"
+#include "qgscurvepolygon.h"
+#include "qgspoint.h"
+#include "qgspolygon.h"
+#include "qgslinestring.h"
+#include "qgsmulticurve.h"
+#include "qgsmultilinestring.h"
+#include "qgsmultipoint.h"
+#include "qgsmultipolygon.h"
+#include "qgsmultisurface.h"
+#include "qgspolyhedralsurface.h"
+#include "qgstriangulatedsurface.h"
+#include "qgstriangle.h"
+#include "qgssfcgalengine.h"
+#include "qgslogger.h"
+
+/**
+ * Wraps SFCGAL geometry object and its QgsAbstractGeometry source
+ * \ingroup core
+ * \class QgsSfcgalGeometry
+ * \brief SfcgalGeometry geometry type.
+ * \since QGIS 3.44
+ */
+class CORE_EXPORT QgsSfcgalGeometry
+{
+  public:
+    //! Constructor for an empty sfcgalGeometry geometry.
+    QgsSfcgalGeometry();
+
+    //! Constructor with SFCGAL shared ptr
+    QgsSfcgalGeometry( sfcgal::shared_geom sfcgalGeom );
+
+    //! Constructor with QgsAbstractGeometry unique_ptr
+    QgsSfcgalGeometry( std::unique_ptr<QgsAbstractGeometry> &qgsGeom );
+
+    //! Constructor with QgsAbstractGeometry ptr
+    QgsSfcgalGeometry( const QgsAbstractGeometry &qgsGeom );
+
+    //! Constructor with QgsAbstractGeometry ptr
+    QgsSfcgalGeometry( const QgsGeometry &qgsGeom );
+
+    //! Constructor from WKT
+    QgsSfcgalGeometry( const QString &wkt, QString *errorMsg = nullptr );
+
+    //! Copy constructor
+    QgsSfcgalGeometry( const QgsSfcgalGeometry &otherGeom );
+
+    //! Returns the underlying SFCGAL geometry
+    sfcgal::shared_geom sfcgalGeometry() const { return mSfcgalGeom; }
+
+    //! Returns the underlying QGIS geometry
+    QgsAbstractGeometry *asQgisGeometry( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Returns type of the geometry as a WKB type (point / linestring / polygon etc.)
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    Qgis::WkbType wkbType( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Returns type of the geometry as String
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    QString geometryType( QString *errorMsg = nullptr ) const SIP_HOLDGIL;
+
+    /**
+     * Clones the geometry by performing a deep copy
+     */
+    QgsSfcgalGeometry *clone() const;
+
+    /**
+     * Creates a new geometry from a WKB byte pointer
+     * \param wkbPtr WKB byte pointer
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    bool fromWkb( QgsConstWkbPtr &wkbPtr, QString *errorMsg = nullptr );
+
+    /**
+     * Export the geometry as WKB
+     *
+     * \param flags argument specifies flags controlling WKB export behavior (since QGIS 3.14).
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    QByteArray asWkb( QgsAbstractGeometry::WkbFlags flags = QgsAbstractGeometry::WkbFlags(), QString *errorMsg = nullptr ) const;
```

Comment:
why not static factory methods for these?

Reply:
I mean instead of forcing construction of an empty QgsSfcgalGeometry and then calling "fromWkb" on it, just make a static factory method:

    `static QgsSfcgalGeometry fromWkb( ... )`